api/V1/recordingUtil.ts

```diff
@@ -143,6 +145,132 @@ export async function tryToMatchRecordingToStation(
   return null;
 }
 
+const THUMBNAIL_MIN_SIZE = 64;
+const THUMBNAIL_PALETTE = "Viridis";
+
+// Gets a raw cptv frame from a recording
+async function getCPTVFrame(recording, frameNumber) {
+  const { CptvDecoder } = await dynamicImportESM("cptv-decoder");
+  const decoder = new CptvDecoder();
+  const fileData = await modelsUtil
+    .openS3()
+    .getObject({
+      Bucket: config.s3.bucket,
+      Key: recording.rawFileKey
+    })
+    .promise()
+    .catch((err) => {
+      return err;
+    });
+  //work around for error in cptv-decoder
+  //best to use createReadStream() from s3 when cptv-decoder has support
+  let data = new Uint8Array(fileData.Body);
+  const meta = await decoder.getBytesMetadata(data);
+  let result = await decoder.initWithLocalCptvFile(data);
+  if (!result) {
+    return;
+  }
+  let finished = false;
+  let currentFrame = 0;
+  let frame;
+  while (!finished) {
+    currentFrame++;
+    frame = await decoder.getNextFrame();
+    finished = frame == null;
+    if (currentFrame == frameNumber) {
+      break;
+    }
+  }
+  decoder.close();
+  return frame;
+}
+
+// Creates and saves a thumbnail for a recording using specified thumbnail info
+async function saveThumbnailInfo(recording, thumbnail) {
+  const frame = await getCPTVFrame(recording, thumbnail.frame_number);
+  const thumb = await createThumbnail(frame, thumbnail);
+
+  let upload = await modelsUtil
+    .openS3()
+    .upload({
+      Bucket: config.s3.bucket,
+      Key: `${recording.rawFileKey}-thumb`,
+      Body: thumb.data,
+      Metadata: thumb.meta
+    })
+    .promise()
+    .catch((err) => {
+      return err;
+    });
+  return upload;
+}
+
+// Create a png thumbnail image  from this frame with thumbnail info
+// Expand the thumbnail region such that it is a square and at least THUMBNAIL_MIN_SIZE
+// width and height
+//render the png in THUMBNAIL_PALETTE
+//returns {data: buffer, meta: metadata about image}
+async function createThumbnail(frame, thumbnail) {
+  const frameMeta = frame.meta.imageData;
+  const res_x = frameMeta.width;
+  const res_y = frameMeta.height;
+
+  const size = Math.max(THUMBNAIL_MIN_SIZE, thumbnail.height, thumbnail.width);
+  const thumbnail_data = new Uint8Array(size * size);
+
+  //dimensions to it is a square with a minimum size of THUMBNAIL_MIN_SIZE
+  const extra_width = (size - thumbnail.width) / 2;
+  thumbnail.x -= Math.ceil(extra_width);
+  thumbnail.x = Math.max(0, thumbnail.x);
+  thumbnail.width = size;
+  if (thumbnail.x + thumbnail.width > res_x) {
+    thumbnail.x = res_x - thumbnail.width;
+  }
+
+  const extra_height = (size - thumbnail.height) / 2;
+  thumbnail.y -= Math.ceil(extra_height);
+  thumbnail.y = Math.max(0, thumbnail.y);
+  thumbnail.height = size;
+  if (thumbnail.y + thumbnail.height > res_y) {
+    thumbnail.y = res_y - thumbnail.height;
+  }
+
+  let frame_start;
+  let thumb_index = 0;
+  for (let i = 0; i < size; i++) {
```

Comment:
I think for a cropped region, you're going to want to get the min/max of the normalisation yourself, since we'd rather normalise for the cropped region than the full frame dynamic range.